### PR TITLE
Add basic deadlock detection when using threadsafe lets and subjects

### PR DESCRIPTION
**Pull request by drcapulet**
Took me a while to trace down why my tests starting locking up after upgrading to 3.3+, figured this would help other people. I'm still working working on integrating the failing repro.

Needs rspec/rspec-support#297.

### CODE REVIEW

**Comment by myronmarston** [comment]
Can you provide an example of a spec that triggers this deadlock condition?

@JoshCheek as the author of the `let`/`subject` threadsafety, do you want to review this?  You're most familiar with the code...

**Comment by drcapulet** [comment]
Spec added - if you remove the raise it'll give you a deadlock error instead.

File: lib/rspec/core/memoized_helpers.rb [hunk 1]
```diff
@@ -153,6 +153,17 @@ def initialize
 
         def fetch_or_store(key)
           @memoized.fetch(key) do # only first access pays for synchronization
+            if @mutex.locked? && !@mutex.owned?
+              raise <<-EOS
+Possible deadlock at:
+  #{CallerFilter.first_non_rspec_line}
+
+This is normally caused when accessing an uninitialized `let` value in a thread
+different than the one calling `subject`. Either use `let!` or disable RSpec
```
Comment (JonRowe) [inline lib/rspec/core/memoized_helpers.rb]: I'd reword this slightly to:

> This is normally caused when accessing an uninitialized `let` value in a different thread. Either use call the let before accessing it or disable RSpec threadsafety with `RSpec.configuration.threadsafe = false`.

Why? `subject` isn't used by everyone (and is just a special `let`) so making it more general should be a good thing, and as `let!` is just the equivalent of `let(:name) {}; before { name }` we have historically discouraged it's use in favour of more direct control.

File: spec/rspec/core/memoized_helpers_spec.rb [hunk 2]
```diff
@@ -1,3 +1,4 @@
+require 'thread'
```
Comment (JonRowe) [inline spec/rspec/core/memoized_helpers_spec.rb]: We should do this in a seperate process to avoid contamination.

File: lib/rspec/core/memoized_helpers.rb [hunk 3]
```diff
@@ -153,6 +153,17 @@ def initialize
 
         def fetch_or_store(key)
           @memoized.fetch(key) do # only first access pays for synchronization
+            if @mutex.locked? && !@mutex.owned?
```
Comment (JoshCheek) [inline lib/rspec/core/memoized_helpers.rb]: For 1.8.7 support, we have [our own Mutex](https://github.com/rspec/rspec-support/blob/5cc75bd37b1bcfb1b9c6c4835d1f1b74fb695fe3/lib/rspec/support/mutex.rb) (see comment for explanation of why), it'll need `locked?` to be defined, which you can get from [here](https://github.com/ruby/ruby/blob/v1_8_7_374/lib/thread.rb#L67-L72). And [our own ReentrantMutex](https://travis-ci.org/rspec/rspec-core/jobs/169634889#L444) which will need `locked?` and `owned?`, both of which should be easy to add. ([Travis confirms](https://travis-ci.org/rspec/rspec-core/jobs/169634889#L444)).

File: lib/rspec/core/memoized_helpers.rb [hunk 4]
```diff
@@ -153,6 +153,17 @@ def initialize
...
+              raise <<-EOS
```
Comment (JoshCheek) [inline lib/rspec/core/memoized_helpers.rb]: * The heredoc is making it difficult to see where this if statement ends (it looks like it might be wrapping the call to `synchronize`). Maybe try inlining? eg `raise <<EOS if @mutex.locked? && !@mutex.owned?`
* Is there an error class we should be using here?
* The dash on the heredoc isn't being used (looks like there's another heredoc in that file doing the same thing, IDK, don't really care, just pointing it out)

File: spec/rspec/core/memoized_helpers_spec.rb [hunk 5]
```diff
@@ -460,6 +462,37 @@ def not_ok?; false; end
...
+              Thread.new do
+                klass.new.bar
+              end.value
+            end
+
+            example do
+              expect { subject }.to raise_error(/deadlock/)
+            end
+          end
+        end
```
Comment (JoshCheek) [inline spec/rspec/core/memoized_helpers_spec.rb]: I think this accomplishes the same thing, and requires a lot less analysis of the test in order to understand it.

```ruby
require 'rspec/autorun'
RSpec.describe do
  let(:lock) do
    Thread.new { lock }.join
  end
  example do
    expect { lock }.to raise_error(/deadlock/)
  end
end

# >> .
# >> 
# >> Finished in 0.10371 seconds (files took 0.20367 seconds to load)
# >> 1 example, 0 failures
# >> 
```

File: spec/rspec/core/memoized_helpers_spec.rb [hunk 6]
```diff
@@ -460,6 +462,37 @@ def not_ok?; false; end
...
+            end
+
+            subject do
+              Thread.new do
+                klass.new.bar
+              end.value
+            end
+
+            example do
+              expect { subject }.to raise_error(/deadlock/)
```
Comment (JoshCheek) [inline spec/rspec/core/memoized_helpers_spec.rb]: Probably do `/Possible deadlock/`, because it looks like were trying to match the error message we'd get without this feature `"No live threads left. Deadlock?"`

File: lib/rspec/core/memoized_helpers.rb (hunk 1 above)
Comment (JoshCheek) [inline lib/rspec/core/memoized_helpers.rb]: +1 on the reasoning
How about this for a message?

> This is caused when one let block is waiting for a thread that accesses another `let` block. The thread gets put to sleep until the first let block finishes, but it never will since it is waiting for the thread.

Comment (JonRowe) [inline lib/rspec/core/memoized_helpers.rb]: Thats great :)

File: spec/rspec/core/memoized_helpers_spec.rb (hunk 5 above)
Comment (JonRowe) [inline spec/rspec/core/memoized_helpers_spec.rb]: We should also use the `in_sub_process` helper and do the require thread within a seperate process.

**Comment by drcapulet** [comment]
Updated, I had to tweak some of the thread ordering because the deadlock detection was throwing an error - it's hard to deterministically decide if we're in a deadlock situation 😞

File: spec/rspec/core/memoized_helpers_spec.rb [hunk 7]
```diff
@@ -402,14 +404,16 @@ def not_ok?; false; end
               if order.current == :second
                 :second_access
               else
-                order.pass_to :second, :resume_on => :sleep
```
Comment (JonRowe) [inline spec/rspec/core/memoized_helpers_spec.rb]: Why are we messing with other specs here? Does this cause a regression elsewhere?

Comment (JonRowe) [inline spec/rspec/core/memoized_helpers_spec.rb]: /cc @JoshCheek ?

Comment (myronmarston) [inline spec/rspec/core/memoized_helpers_spec.rb]: With your change will we start giving this new "probable deadlock" error for non-deadlock situations where threads where interleaved? That seems problematic and like it may break people's specs.

Comment (drcapulet) [inline spec/rspec/core/memoized_helpers_spec.rb]: You'll only see this "possible deadlock" error for non-deadlock situations where threads are interleaved when multiple threads are accessing the same uninitialized `let` (or a `let` inside of a `subject` from another thread), which seems like it'd be rare but obviously up to y'all.

**Comment by drcapulet** [comment]
Ping @JoshCheek / @JonRowe / @myronmarston, anything need to be changed to get this merged?

**Comment by myronmarston** [comment]
Given that this change causes a "possible deadlock" error to be raised in situations that (a) worked previously and (b) are not a deadlock, I lean against merging this.  I see the usefulness of your change for situations where the are deadlocks, but given how widely used RSpec is (we have 50 million+ downloads....) I don't think we can assume that users won't get the error in a situation that previously worked.

I'm curious what @JoshCheek thinks, though.  Or is there a way to provide this without breaking existing specs such as the ones you had to change here?

**Comment by JonRowe** [comment]
Yeah I'm also uncomfortable with the change in behaviour

**Comment by drcapulet** [comment]
Understandable! Fortunately MRI detects deadlock situations but not JRuby, hence the source of this PR.

Would something more complex that detected dependency cycles be more acceptable?

**Comment by myronmarston** [comment]
> Understandable! Fortunately MRI detects deadlock situations but not JRuby, hence the source of this PR.

Ah, so are you really just trying to improve the situation when you run into this on JRuby?  If so, would it make sense to ask the JRuby team to improve its deadlock detection support to be more like MRI's?

**Comment by JoshCheek** [comment]
My understanding changed about 3x while writing this, I'd reread it to make sure it's coherent with my final one, but I really need to go walk the dogs and shower and smoke and read my texts, so some bits might be a little incongruent.


High level explanation
----------------------

So, looking at that first test that was changed, the point of the test is that turning on threadsafety has the behaviour that the "first thread to access determines the return value", the reason for this is because without threadsafety, the last thread that returned from the `let` block would overwrite the value that earlier threads had calculated. This test recreates that situation, but the changes mean that it no longer tests this.

**The old non-threadsafe behaviour**: thread1 enters, thread2 enters, thread1 leaves and sets memoized value, thread2 leaves and overrides memoized value. When we call the let block, it is returning different objects.

**What this test does** to verify the first thread to access determines the return value: thread1 enters, thread2 tries to enter but gets put to sleep, thread1 leaves and sets memoized value, thread2 wakes and uses thread1's memoized value

**What the modified test does** which does not verify it: thread1 enters, thread1 leaves and sets the memoized value, thread2 uses thread1's memoized value

As I've attempted to think of what to do from here, I've realized the reason these needed to be changed is probably that the "possible deadlock" test checks for a collision, not a deadlock collision. Collisions are fine, clearly we expect them as we engineer them in this test. This probably also explains why there are JRuby issues: JRuby allows threads to run in parallel where MRI has the GIL to prevent it. Maybe with real parallelism, the `let` block threadsafety is being used, but this catches it and calls it a deadlock. I talk a bit about where I'd go from here down at the bottom.


Specific explanation
--------------------

"thread1" and "thread2" from the explanation above will be called "first" and `:second` here, in order to match the code.

First, we'll walk through the [old](https://github.com/drcapulet/rspec-core/blob/16888f92347036e96ef42d7b0e4449929778e877/spec/rspec/core/memoized_helpers_spec.rb#L396-L415) code.

* We call `memoized_value` from first (L412)
* That goes into the let block, the current thread is first, not `:second`, so it goes to the false branch (L402)
* First goes to sleep until `:second` resumes (L405)
* `:second` is started. If it let were not threadsafe, it would go into the let block and to the true branch, returning `:second_access` However, b/c `let` is threadsafe, `let` puts `:second` to sleep (**this is probably where deadlock detection is hitting an issue, because they are now both sleeping**) (L411)
* Now that `:second` is asleep, first wakes up (because it passed `:resume_on => :sleep`), and returns `:first_access` (L406, L412)
* Now that first has left the let block, `:second` wakes up and does not need to enter.
* It receives the `:first_access`, that first calculated, makes its assertion, and exits (L411)

Now, we'll walk through the [changed](https://github.com/drcapulet/rspec-core/blob/9e65fc16d7f73a1ba062a5bd9cb65edaf4ceb206/spec/rspec/core/memoized_helpers_spec.rb#L398-L419) code.

* We call `memoized_value` from first (L415)
* That goes into the let block
* The current thread is first, not `:second` (L404), so it goes to the false branch and returns `:first_access` (L407, L415)
* At this point, the memoized value is calculated.
* First then goes to sleep until `:second` exits (L416)
* `:second` starts, accesses the `let` block which is already memoized so it never attempts to enter the `let` block, it asserts that it saw first's memoized value and exits (L413)
* First resumes and the test passes. (L416)


The problem with the feature
----------------------------

So, this test puts these threads into the situation the code detects, however, there is no deadlock here, because there is at least one thread still running: `ThreadOrder`'s worker is processing its event queue. So the current implementation of the PR is incorrect.

Here's another way to see it: Spawn a bunch of threads, have them call a let block with a sleep statement in it. They should all fall asleep until the first one finishes (there is no deadlock it's just slow). But the current way this is implemented misidentifies it as a deadlock. I ran this just to make sure b/c it's going to be a pretty big hurdle to overcome: https://gist.github.com/JoshCheek/43c8735ccfceb48d370db6a9be66cb53

So what really needs to happen is a smarter deadlock detection algorithm, currently I think it's just detecting concurrent access and raising, but the goal of threadsafety is accommodate concurrent access.


Idea 1: Figure out how Ruby detects it
--------------------------------------

Maybe the real question is whether the owner is asleep waiting for the current thread to do something. Obviously there is some way to know because Ruby can tell, maybe start by looking at [how they do it](https://github.com/ruby/ruby/blob/121ce4ac90ee4aff2ee3c9b84cf46b78a58af123/thread.c#L4898-L4934).

```sh
$ ruby -e 'Queue.new.shift'
-e:1:in `pop': No live threads left. Deadlock? (fatal)
	from -e:1:in `<main>'

$ ruby -e sleep # sleeping forever is not seen as a deadlock
```


Idea 2: Piggyback on Ruby's deadlock detection
----------------------------------------------

It just occurred to me that we can piggy back on Ruby's deadlock detection, which will save us from having to anticipate it in advance. When we detect that the error has occurred, print some debugging info to stderr and let Ruby finish killing the process.

Ensure blocks still run:

```ruby
begin
  Queue.new.shift # ~> fatal: No live threads left. Deadlock?
ensure
  puts "executed despite the deadlock"
end
```

At exit hooks still work:

```ruby
at_exit { puts "some error checking code or w/e" }
Queue.new.shift # ~> fatal: No live threads left. Deadlock?

# >> some error checking code or w/e
```

We can technically even rescue it and do w/e (least viable unless we immediately raise again):

```ruby
begin
  Queue.new.shift
rescue Object
  $!                          # => #<fatal: No live threads left. Deadlock?>
  $!.class.new "custom error" # => #<fatal: custom error>
end
```

Idea 3: Abandonment
-------------------

I think that idea 2 will actually work and it probably makes sense to do (I didn't see the struggle to find the problem test, but could see it being a PITA). But if it turns out to not be viable either, then lets keep in mind that Ruby's default error message here is `No live threads left. Deadlock?`, which honestly isn't that bad.

**Comment by myronmarston** [comment]
Thanks for the thorough response, @JoshCheek!

Given what you've explained, I think it's clear that the implementation in this PR is problematic and should not be merged.  I'm going to close this PR.  @drcapulet if you want to use an alternate solution like what @JoshCheek suggested above, feel free to open another PR.

We should probably also revert rspec/rspec-support#297 since we don't actually need those changes.

**Comment by JoshCheek** [comment]
I think I'm going to advocate @drcapulet open a new one that adds `ensure` or `rescue` to the method with the deadlock and then prints the error message to stderr. I think it will a simple solution that will be more reliable than the implementation here and I think it will have value for users b/c it will tell them where the problem is, and we could add a bit more info about what to do in that situation so that they are a bit more empowered themselves. 

Most of the work is done at this point:

* Probably can use the same test
* I've got a lot more of the nuances loaded into my brain to give useful feedback
* It can be played with in isolation via `Queue.new.shift`
* Examples to get started with are in "Idea 2" above
* It'll mostly be about correctly detecting the error, which you can probably do by checking the error's class and message:

  ```ruby
  begin
    if fork                          # => 18920, nil
      raise 'some irrelevant error'
    else
      Queue.new.shift                # ~> fatal: No live threads left. Deadlock?
    end
  ensure
    $!.class.name == 'fatal'         # => false, true
    $!.message =~ /deadlock/i        # => nil,   22
  end
  ```

**Comment by JonRowe** [comment]
@joshcheek I don't think that fixes the issue @drcapulet's issue, I think on JRuby the issue is the deadlock actually happens rather than a nice error on MRI (as MRI is actually faking the threads and knows it's happening).

My hypothetical solution to this (not all deadlock issues but at least this one), is we could keep a stack of accessed lets during the assignment process, and if we attempt to access the same one from inside the same thread (but only the same thread so as not to recause the non threadsafe problem), raise an error.
